### PR TITLE
feat: add per-head fp8 kv attention calibration

### LINKED ISSUE #2026: [Feature]: Extend the FP8 KV/Attn Calibration to Per-head
### Feature Description

Extend FP8 KV/Attention calibration from per-tensor scaling to per-head scaling.

### Context

We previously added support for collecting per-tensor scales for FP8 KV/Attention calibration, including usage examples and unit tests:

https://github.com/intel/auto-round/blob/5e586bcbc0b9407666576287768ca7ca09365652/test/test_cpu/export/test_llmc_format.py#L131-L152

This feature extends the calibration granularity to per-head scaling, meaning each attention head has its own scale. Compared with per-tensor scaling, per-head scaling can better capture head-wise activation ranges and may improve accuracy while keeping inference overhead minimal, assuming the target backend supports per-head scales efficiently.

### Definition of Done

* Support per-head FP8 KV/Attention calibration for:

  - [ ] MHA
  - [ ] GQA
  - [ ] MLA
* Support export formats:

  - [ ] LLMC
  - [ ] AutoRound
* For the LLMC format, validate accuracy on the vLLM side using the Triton backend.

### Additional Context

N/A

### PR DESCRIPTION
**Pull request by zeel2104**
Resolve https://github.com/intel/auto-round/issues/2026
## Description

Adds per-head FP8 KV/attention calibration support on top of the existing per-tensor static FP8 paths.

Main changes:
- Adds `static_kv_granularity` and `static_attention_granularity` options with `"tensor"` as the default and `"head"` as the new per-head mode.
- Collects per-head `k_scale`, `v_scale`, and `q_scale` for attention tensors shaped as `[batch, heads, seq, head_dim]`.
- Exports LLMC metadata with `attn_head` strategy for per-head KV/attention scales.
- Preserves existing per-tensor behavior by default.
- Adds tests for helper MHA/GQA scale shapes, LLMC export, and AutoRound export.
- Fixes a Windows-only test file-lock issue by closing `safe_open` handles before reusing the output path.

## Type of Change

New feature

## Related Issues

Relates to #2026

## Checklist Before Submitting

- [x] My code has been tested locally.
- [ ] Documentation has been updated as needed.
- [x] New or updated tests are included where applicable.
- [ ] The CUDA CI has passed. You can trigger it by commenting `/azp run Unit-Test-CUDA-AutoRound`.

Tested locally:
- `python -m pytest test/test_cpu/test_experimental_fp8_granularity.py -q`
- `python -m pytest test/test_cpu/export/test_llmc_format.py -q`
- `python -m pytest test/test_cpu/export/test_export.py -k "per_head or static_afp8_export or static_fp8_attn" -q`
- `python -m pytest test/test_cpu/quantization/test_mxfp_nvfp.py -k "fp8_kv_attn" -q`

Note:
- `test/test_cpu/integrations/test_inc_integration.py` was not run because `neural_compressor` is not installed locally.

### CODE REVIEW

**Comment by yiliu30** [comment]
Hi @zeel2104, thanks for the contribution! Overall, it looks good to me. Could you please rebase the changes onto the latest main branch and resolve the CI issues?

We also need to quantize the model and validate inference in vLLM using the Triton backend.

**Comment by yiliu30** [comment]
Hi @zeel2104, are you still working on this PR? We're targeting this feature for the current release. If you're tied up with other work, please let me know. Thanks!

**Comment by zeel2104** [comment]
Hi @yiliu30, sorry for late reply. I rebased this PR onto the latest `main` and pushed the updated branch.

I also resolved the LLMC exporter conflict by keeping the newer top-level `attention_input_activations` schema and applying the per-head `attn_head` strategy there.

Local validation after rebase:
- `python -m pytest test/test_cpu/test_experimental_fp8_granularity.py -q`
- `python -m pytest test/test_cpu/export/test_llmc_format.py -k "kv_config or attention_config" -q --tb=short`
- `python -m pytest test/test_cpu/export/test_export.py -k "per_head or static_afp8_export or static_fp8_attn" -q --tb=short`

I do not have vLLM/Triton available locally, so I’ll need CI or a CUDA environment for the vLLM Triton inference validation.

**Review by yiliu30 — CHANGES_REQUESTED** [review]
Findings

- [P1] auto_round/experimental/utils.py:69: The new per-head FP8 QDQ path casts directly through `torch.float8_e4m3fn` instead of using the project’s STE helper. Tensor granularity uses `float8_e4m3fn_ste()` in `auto_round/data_type/fp8.py`, which preserves gradient flow during optimized calibration. With `static_*_granularity="head"` and `iters > 0`, this new path can cut or mishandle gradients through Q/K/V, so per-head mode is not equivalent to the existing static FP8 behavior. Please route the per-head cast through the same STE helper and add a gradient-flow test.

**Comment by yiliu30** [comment]
> Hi @yiliu30, sorry for late reply. I rebased this PR onto the latest `main` and pushed the updated branch.
> 
> I also resolved the LLMC exporter conflict by keeping the newer top-level `attention_input_activations` schema and applying the per-head `attn_head` strategy there.
> 
> Local validation after rebase:
> 
> * `python -m pytest test/test_cpu/test_experimental_fp8_granularity.py -q`
> * `python -m pytest test/test_cpu/export/test_llmc_format.py -k "kv_config or attention_config" -q --tb=short`
> * `python -m pytest test/test_cpu/export/test_export.py -k "per_head or static_afp8_export or static_fp8_attn" -q --tb=short`
> 
> I do not have vLLM/Triton available locally, so I’ll need CI or a CUDA environment for the vLLM Triton inference validation.

Hi @zeel2104 it is fine, could you quantize the Qwen3-8B and upload it the hugging face. I can help verify it.

File: auto_round/cli/parser.py [hunk 1]
```diff
@@ -184,13 +184,27 @@ def build_quantize_parser(*, prog: str = "auto_round quantize") -> argparse.Argu
         choices=["fp8", "float8_e4m3fn"],
         help="Static KV-cache quantization data type.",
     )
+    rt.add_argument(
+        "--static_kv_granularity",
+        default="tensor",
+        type=str,
+        choices=["tensor", "head"],
```
Comment (wenhuach21) [inline auto_round/cli/parser.py:191]: We already have too many CLI arguments. Is there a way to group all the KV-related options into a kv_scheme and provide some predefined presets to cover the common use cases?

File: auto_round/context/compress.py [hunk 2]
```diff
@@ -58,6 +60,8 @@ def __init__(
         self.is_immediate_saving = is_immediate_saving
         self.static_kv_dtype = static_kv_dtype
         self.static_attention_dtype = static_attention_dtype
+        self.static_kv_granularity = static_kv_granularity
```
Comment (wenhuach21) [inline auto_round/context/compress.py:63]: Personally, I think KV-related options belong in the **scheme** context rather than the **compression** context.

@n1ck-guo, could you refine the definition of each context and clearly explain its purpose? That would make it much easier for developers to understand which context a new option should belong to.

File: auto_round/cli/parser.py (hunk 1 above)
Comment (yiliu30) [reply auto_round/cli/parser.py:191]: Yes, I’m also thinking about refining the KV-cache and attention quantization configs into dedicated schemes, something like:

```json
"kv_cache_scheme": {
    "dynamic": false,
    "group_size": null,
    "num_bits": 8,
    "strategy": "tensor"
    ...
}
```

I’ll prepare an RFC for this later. For this PR, though, I think the current functionality looks good.

File: test/unit/test_cpu/export/test_export.py [hunk 3]
```diff
@@ -319,7 +359,37 @@ def test_static_fp8_attn(self):
...
+            self.tokenizer,
+            iters=0,
+            nsamples=2,
+            seqlen=2,
+            scheme="FP8_STATIC",
+            static_attention_dtype="fp8",
+            static_attention_granularity="head",
+        )
+        _, quantized_model_path = autoround.quantize_and_save(output_dir=self.save_dir, format="auto_round")
+        f = safe_open(os.path.join(quantized_model_path, "model.safetensors"), framework="pt")
```
Comment (lvliang-intel) [inline test/unit/test_cpu/export/test_export.py:386]: `safe_open` without a `with` statement, inconsistent with the
context-manager pattern applied to the other tests in this commit.

File: auto_round/experimental/utils.py [hunk 4]
```diff
@@ -32,6 +48,37 @@ def per_tensor_fp8_qdq(
     return qdq_tensor, scale
 
 
+def per_head_fp8_qdq(tensor: torch.Tensor, tensor_max: None | torch.Tensor = None) -> tuple[torch.Tensor, torch.Tensor]:
```
Comment (lvliang-intel) [inline auto_round/experimental/utils.py:51]: Consider adding a shape guard for tensor_max

File: auto_round/experimental/utils.py [hunk 5]
```diff
@@ -41,7 +88,10 @@ def update_parameter_data(module: torch.nn.Module, new_val: torch.Tensor, name:
     if hasattr(module, name):
         param = getattr(module, name)
         if isinstance(param, torch.nn.Parameter):
-            param.data.copy_(new_val)
+            if param.shape == new_val.shape:
+                param.data.copy_(new_val)
+            else:
+                setattr(module, name, torch.nn.Parameter(new_val))
```
Comment (lvliang-intel) [inline auto_round/experimental/utils.py:105]: Consider at least logging a warning when the replacement path is taken, so the
behavior change is visible.

File: test/unit/test_cpu/export/test_export.py (hunk 3 above)
Comment (yiliu30) [reply test/unit/test_cpu/export/test_export.py:386]: fixed

File: auto_round/experimental/utils.py (hunk 4 above)
Comment (yiliu30) [reply auto_round/experimental/utils.py:51]: udpated

File: auto_round/experimental/utils.py (hunk 5 above)
Comment (yiliu30) [reply auto_round/experimental/utils.py:105]: added

**Review by yiliu30 — APPROVED** [review]

**Review by lvliang-intel — APPROVED** [review]
LGTM
